### PR TITLE
Automatically delete empty communities

### PR DESCRIPTION
**Pull request by Changaco**
#3114

### CODE REVIEW

**Comment by Changaco** [comment]
Diff of the `upsert_community` function:

```diff
-CREATE FUNCTION upsert_community() RETURNS trigger AS $$
+CREATE OR REPLACE FUNCTION upsert_community() RETURNS trigger AS $$
     DECLARE
         is_member boolean;
+        delta int = CASE WHEN NEW.is_member THEN 1 ELSE -1 END;
     BEGIN
         IF (SELECT is_suspicious FROM participants WHERE id = NEW.participant) THEN
             RETURN NULL;
@@ -18,21 +19,27 @@
         END IF;
         LOOP
             UPDATE communities
-               SET nmembers = nmembers + (CASE WHEN NEW.is_member THEN 1 ELSE -1 END)
-             WHERE slug = NEW.slug;
+               SET nmembers = nmembers + delta
+             WHERE slug = NEW.slug
+               AND nmembers + delta > 0;
             EXIT WHEN FOUND;
-            BEGIN
-                INSERT INTO communities
-                     VALUES (NEW.slug, NEW.name, 1, NEW.ctime);
-            EXCEPTION
-                WHEN unique_violation THEN
-                    IF (CONSTRAINT_NAME = 'communities_slug_pkey') THEN
-                        CONTINUE; -- Try again
-                    ELSE
-                        RAISE;
-                    END IF;
-            END;
-            EXIT;
+            IF (NEW.is_member) THEN
+                BEGIN
+                    INSERT INTO communities
+                         VALUES (NEW.slug, NEW.name, 1, NEW.ctime);
+                EXCEPTION
+                    WHEN unique_violation THEN
+                        IF (CONSTRAINT_NAME = 'communities_slug_pkey') THEN
+                            CONTINUE; -- Try again
+                        ELSE
+                            RAISE;
+                        END IF;
+                END;
+                EXIT;
+            ELSE
+                DELETE FROM communities WHERE slug = NEW.slug AND nmembers = 1;
+                EXIT WHEN FOUND;
+            END IF;
         END LOOP;
         RETURN NEW;
     END;
```

File: branch.sql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                    WHEN unique_violation THEN
+                        IF (CONSTRAINT_NAME = 'communities_slug_pkey') THEN
+                            CONTINUE; -- Try again
+                        ELSE
+                            RAISE;
+                        END IF;
+                END;
+                EXIT;
+            ELSE
+                DELETE FROM communities WHERE slug = NEW.slug AND nmembers = 1;
```
Comment (whit537) [inline branch.sql]: `communities` is an INSERT-only table. Why should we start DELETEing from it here?

**Comment by whit537** [comment]
I don't think this is the right solution to #3114, because the `communities` table is designed to be `INSERT`-only. There's another way to implement #3114 without `DELETE`ing from `communities`.

**Comment by whit537** [comment]
[IRC](https://botbot.me/freenode/gratipay/2015-02-05/?msg=31341990&page=2)

**Pull request merged**